generator/templates/structs.cpp.j2

```diff
@@ -0,0 +1,167 @@
+    {%- for sub_frame in msg.frame -%}
+        {{ get_message_ranges(sub_frame, msg_names + [sub_frame.name]) }}
+    {%- endfor -%}
+{%- else -%}
+     std::make_pair(1, 0),
+{%- endif -%}
+{%- endmacro -%}
+
+{%- macro get_frame_ranges(msg, msg_names, cnt = 0) -%}
+{%- if msg.frame is defined -%}
+    {{ get_msg_len(msg) }},
+    {%- for sub_frame in msg.frame -%}
+        {{ get_frame_ranges(sub_frame, msg_names + [sub_frame.name], cnt + 1) }}
+    {%- endfor -%}
+{%- else -%}
+     0,
+{%- endif -%}
+{%- endmacro -%}
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <chrono>
+#include "static.hpp"
+#include "bus.hpp"
+#include "structs.hpp"
+
+namespace CANlib {
+
+void send(Clock::time_point &last_sent_, Message_T* msg) {
+    if (Clock::now() - last_sent_ > msg->get_period()) {
+        last_sent_ = Clock::now();
+        Frame frame;
+        msg->unpack(frame);
+        TransmitFrame(frame, msg->get_bus_name());
+    }
+}
+
+void send(Message_T* msg) {
+    Frame frame;
+    msg->unpack(frame);
+    TransmitFrame(frame, msg->get_bus_name());
+}
+
+{%- for bus in can.bus %}
+namespace {{bus.name}} {
+{% for msg in bus.frame -%}
+{{ struct_declarations(bus, msg, msg.name) }}
+{%- endfor -%}
+
+extern const Message_Type_T message_types[{{get_msg_len(bus)}}] = {
+    Message_Type_T::UNKNOWN_MSG,
+    {% for msg in bus.frame -%}
+        {{ get_message_type(bus, msg, [msg.name]) }}
+    {%- endfor -%}
+};
+
+}
+
+const uint32_t {{bus.name}}_keys[] = {
+    0,
+    {% for msg in bus.frame -%}
+        {{ get_keys(msg, [bus.name, msg.name]) }}
+    {%- endfor -%}
+};
+
+const std::pair<uint32_t, uint32_t> {{bus.name}}_frame_id_range[] = {
+    std::make_pair(1,0),
+    {% for msg in bus.frame -%}
+        {{ get_message_ranges(msg, [bus.name, msg.name]) }}
+    {%- endfor -%}
+};
+
+const uint32_t {{bus.name}}_frame_len[] = {
+    0,
+    {% for msg in bus.frame -%}
+        {{ get_frame_ranges(msg, [bus.name, msg.name]) }}
+    {%- endfor -%}
+};
+
+Message_T* {{bus.name}}_messages[] = {
+    nullptr,
+    {%- for msg in bus.frame -%}
+        {{ get_messages(msg, ['CANlib', bus.name, msg.name]) }}
+    {%- endfor -%}
+};
+{%- endfor -%}
+
+const uint32_t* keys[static_cast<int>(Bus_Names_T::INVALID_NAME)] = {
+{%- for bus in can.bus -%}
+{{bus.name}}_keys,
+{%- endfor -%}
+};
+
+const std::pair<uint32_t, uint32_t>* frame_id_range[static_cast<int>(Bus_Names_T::INVALID_NAME)] = {
+{%- for bus in can.bus -%}
+{{bus.name}}_frame_id_range,
+{%- endfor -%}
+};
+
+const uint32_t* frame_len[static_cast<int>(Bus_Names_T::INVALID_NAME)] = {
+{%- for bus in can.bus -%}
+{{bus.name}}_frame_len,
+{%- endfor -%}
+};
+
+extern const size_t can_size[static_cast<int>(Bus_Names_T::INVALID_NAME)] = {
+{%- for bus in can.bus -%}
+{{ get_msg_len(bus) }},
+{%- endfor -%}
+};
+
+Message_T** messages[static_cast<int>(Bus_Names_T::INVALID_NAME)] = {
+{%- for bus in can.bus -%}
+{{bus.name}}_messages,
+{%- endfor -%}
+
+};
+}
```

Comment:
Wait, so if I want a particular input struct, how exactly does that work?

Reply:
type `extern CANlib::J::CC::AAA AAA_input;` and use `AAA_input`, alternatitely you can do extern `CANlib::Message** messages; ` and use `messages[0][static_cast<int>(map1::MessageType::A)]`